### PR TITLE
Feat/add optional policies

### LINKED ISSUE #33: Option to withdraw acceptance of terms (optional consents).
## Feature Request
An option to add optional terms would be useful. E.g. Marketing consents, etc.
Currently this extension requires acceptance of all terms and does not allow you to withdraw from them.

- In the admin panel add a checkbox "This terms are optional".
- If the terms are optional then the extension should not require it
- User in profile settings can withdraw consent

### PR DESCRIPTION
**Pull request by AngryDragonite**
**Resolves #33**

**Changes proposed in this pull request:**

First of all,  I added migrations that create columns: `optional` (boolean) and `additional_info` (JSON) inside `fof_terms_policies` table. Table `fof-terms-policy-user`  has been enriched with `is_accepted` (boolean) column.

Modifications inside `extend.php`:
- added styles to forum
- added controller that handles rejection of an (optional) policy,
- added relationship _fofTermsPoliciesState_ that allows to check which policies were accepted,
- removed _Registered_ event listener, its functionality is now being handled inside `RegisterMiddleware.php`.

Modifications inside `Policy.php`, `UserPolicyData.php`, `PolicyStoreController.php`, `PolicyRepository.php`, `RegisterPolicyValidator.php`:

- generally speaking, added functionality that would integrate backend with new columns introduced in migrations,
- validator now forces users to accept only non-optional policies.

Modifications inside `RegisterMiddleware.php`:
- added functionality that handles storing the state of user's policies (wheter a policy was accepted, or not) to database.

New backend files:

- `PolicyDeclineController.php` - as the name suggests. Basically, it is "1:1 copy" of  `PolicyAcceptController.php`, however with opposite behaviour

Now lets take a look at what changed in the frontend.

Modifications inside `Policy.js`:
- added `optional` and `additional_info` fields to the  **Policy** model.

Modifications inside `PolicyEdit.js`:
- added new item to **fields** method, which allows to choose, wheter specified **Policy** should be optional, or not,
- synced `initNewField` method with `Policy` model changes,
- modified `deletePolicy` method - during developing I encountered a bug - when deleting a `Policy`, its field is, of course deleted, but also `savePolicy` method is being called. It was happening possibly because when clicking _Delete policy_ button, `onsubmit` method is being called.

Modifications inside `AcceptPoliciesModal.js`:
- added functionality that handles rejecting an optional policy.

Modifications inside `forum/index.js':
- instantiating new `addManagePoliciesOption` component.

Modifications inside `admin/components/index.js`:
- exporting **new** `ExtensionData` component, so developers working on integrating **fof/terms** with other extensions could access this component

New frontend files:
- `ExtensionData.js` - a component that allows developers to integrate **fof/terms** with other extensions by saving their data into `fof-terms-policies` table. Recipe on how to use it is descriped in `README.md` in _For developers_ section.
-  `addManagePoliciesOption.js` - a component that allows users to opt in/out of consents in their profile settings.

Other changes:
- `admin.less` & **new** `forum.less`  - added some styling for components,
-  `en.yml` - added translations.
- `README.md` - added section _For developers_, explaining step-by-step how to use `ExtensionData` component, and therefore integrate `fof/terms` with other extensions.

**Reviewers should focus on:**

I would definitely want to know, if my instructions on how to use `ExtensionData` in `README.md` are specific, exact and free of ambiguities. Every feedback on what in my code can be fixed, rewritten so it is easier to read, or is simply better is greatly welcomed :)

**Screenshot**

image
&ensp; &ensp; &ensp; &ensp; &ensp; &ensp; _admin panel - select policy as optional_
\
\
image
&ensp; &ensp; &ensp; &ensp; &ensp; &ensp; _user settings - opt in/out of consents_

**Confirmed**

- [x] Frontend changes: tested on a local Flarum installation.
- [x] Backend changes: tests are green (run `composer test`).

**Required changes:**

- [ ] Related [Flarum core extension PR's](https://github.com/flarum/core/pulls): (Omit this section if irrelevant)

### CODE REVIEW

File: js/src/forum/components/addManagePoliciesOption.js
Comment (DavideIadeluca) [inline js/src/forum/components/addManagePoliciesOption.js:38]: If a URL is provided, this should be opened in a new tab with `target="_blank"`. If no URL is provided by the admin, I think it should be just a regular text. But I'm curious what others are thinking about this...

**Review by DavideIadeluca — CHANGES_REQUESTED** [review]
Thanks @AngryDragonite for your contribution! Just some minor things to iron out :)

File: extend.php
Comment (AngryDragonite) [reply extend.php:1]: Hello, I couldn't access, edit or save some files without using `sudo` all the time, so I decided to change permissions.

File: resources/locale/en.yml
Comment (AngryDragonite) [reply resources/locale/en.yml]: Great idea! Committed changes.

File: resources/locale/en.yml
Comment (AngryDragonite) [reply resources/locale/en.yml]: I definetely agree :) This approach provides much more clarity. Changes committed.

File: migrations/2024_11_4_07_edit_terms_policies.php
Comment (AngryDragonite) [reply migrations/2024_11_4_07_edit_terms_policies.php:1]: Changes commited.

File: extend.php
Comment (AngryDragonite) [reply extend.php:1]: Reverted file permissions, commited changes

File: js/src/forum/components/addManagePoliciesOption.js
Comment (AngryDragonite) [reply js/src/forum/components/addManagePoliciesOption.js]: Proposed changes were committed.

File: migrations/2024_11_4_124021_edit_terms_policies.php
Comment (DavideIadeluca) [inline migrations/2024_11_4_124021_edit_terms_policies.php:1]: Rename it to: `2024_04_11_01_edit_terms_policies.php`

explanation: 
- `2024` equals the year the migration was written
- `04` equals the month the migration was written
- `11` equals the day of the month the migration was written
- `01` is used as an order (especially required if multiple migrations are written in a single day, but good to include anyways). The official docs handle this order differently, but having just `01` is okay in this case

File: migrations/2024_11_4_124110_edit_terms_policy_user.php
Comment (DavideIadeluca) [inline migrations/2024_11_4_124110_edit_terms_policy_user.php:1]: Rename it to: `2024_04_11_01_edit_terms_policy_user.php`

File: migrations/2024_23_04_124250_edit_terms_policies.php
Comment (DavideIadeluca) [inline migrations/2024_23_04_124250_edit_terms_policies.php:1]: Rename it to: `2024_04_24_01__edit_terms_policies.php`

**Review by DavideIadeluca — CHANGES_REQUESTED** [review]
Thanks @AngryDragonite for the changes. The migrations still need a look, otherwise this PR from my side is good for a merge

**Comment by AngryDragonite** [comment]
All requested changes committed.

**Review by DavideIadeluca — APPROVED** [review]

**Review by imorland — CHANGES_REQUESTED** [review]
Thanks for the PR @AngryDragonite and sorry it's taken us so long to look at this.

Overall I think it's okay, but I'd like to give it a spin locally before merging. Please will you remove the committed `dist` folder from the PR? This is not needed here as these files will be built automatically upon merge, and only cause merge conflicts at this stage.

You can do this easily: `git reset HEAD js/dist` and then push that change :)

**Comment by imorland** [comment]
Thank you @AngryDragonite - however there are now more files in `js/dist`. Please remove the `dist` folder from your commits to resolve the conflicts

**Comment by AngryDragonite** [comment]
@imorland Sorry for confusion, I followed steps that you advised, but somehow it got worse, possibly because I messed up something. I'm trying to fix it now

**Comment by imorland** [comment]
@AngryDragonite thanks for your efforts, unfortunately you're now deleting the `dist` folder contents, rather than simply omitting any changes to them. [This](https://itsnotbugitsfeature.com/2020/01/24/removing-a-specific-file-from-a-pull-request/) should help you ;)

**Comment by rafaucau** [comment]
@imorland There are changes to the dist folder on the master branch since this PR. Perhaps this is causing the issues?
https://github.com/FriendsOfFlarum/terms/commits/master/js/dist?since=2024-04-28&until=2024-06-16

I tried to help @AngryDragonite, but still the files are in this Pull Request.

File: js/src/forum/components/addManagePoliciesOption.js
Comment (DavideIadeluca) [inline js/src/forum/components/addManagePoliciesOption.js:1]: I just saw that this file isn't yet formatted correctly. Please run `yarn format`. Then we should be good to go! :)

Comment (DavideIadeluca) [reply js/src/forum/components/addManagePoliciesOption.js:1]: @rafaucau @AngryDragonite

Comment (rafaucau) [reply js/src/forum/components/addManagePoliciesOption.js:1]: On the branch master Prettier is updated to version 3. I will need to cherry-pick commits to get this formatted according to the current configuration

Comment (rafaucau) [reply js/src/forum/components/addManagePoliciesOption.js:1]: I went around it. I manually updated prettier (still using npm) and did not commit the change.

**Review by DavideIadeluca — APPROVED** [review]

**Review by imorland — CHANGES_REQUESTED** [review]

File: src/Middlewares/RegisterMiddleware.php
Comment (rafaucau) [reply src/Middlewares/RegisterMiddleware.php:90]: This is handled earlier above. We need to handle this to read response

File: src/Middlewares/RegisterMiddleware.php
Comment (rafaucau) [reply src/Middlewares/RegisterMiddleware.php]: Removed

File: src/Middlewares/RegisterMiddleware.php
Comment (rafaucau) [reply src/Middlewares/RegisterMiddleware.php]: Done

File: js/src/forum/components/addManagePoliciesOption.js
Comment (rafaucau) [reply js/src/forum/components/addManagePoliciesOption.js]: Done

File: migrations/2024_04_24_01_edit_terms_policies.php
Comment (rafaucau) [reply migrations/2024_04_24_01_edit_terms_policies.php:11]: In this case, we can check the database version and, if it is too old, set the column as text. A column of type text will not cause problems because JSON is text and Eloquent will still cast the model attribute to an array in the same way.

@imorland Do you know how to execute a custom SQL query in Flarum? Unfortunately, the DB facade doesn't work.

```php
<?php

use Illuminate\Database\Schema\Blueprint;
use Illuminate\Database\Schema\Builder;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Str;

return [
    'up' => function (Builder $schema) {
        $schema->table('fof_terms_policies', function (Blueprint $table) {
            // JSON columns are supported on MariaDB 10.2+, MySQL 5.7+
            $requiredVersion = '5.7.8';
            $mySqlVersion = DB::select('SELECT VERSION() as version')[0]->version;

            if(Str::contains($mySqlVersion,'MariaDB')){
                $requiredVersion = '10.2.7';
            }

            if(version_compare($mySqlVersion, $requiredVersion,'>=')){
                $table->json('additional_info')->nullable();
            } else {
              $table->text('additional_info')->nullable();
            }
        });
    },
    'down' => function (Builder $schema) {
        $schema->table('fof_terms_policies', function (Blueprint $table) {
            $table->dropColumn('additional_info');
        });
    },
];
```

Comment (rafaucau) [reply migrations/2024_04_24_01_edit_terms_policies.php]: I've submitted the change. The `additional_info` column will now be of type `LONGTEXT` instead of `JSON`. The effect will be the same, but there won't be native JSON validation in the database.

**Comment by rafaucau** [comment]
ping @imorland

**Comment by rafaucau** [comment]
ping @imorland @DavideIadeluca

**Comment by rafaucau** [comment]
Maybe someone else from the FOF team could review this PR? Like, for example, @dsevillamartin

**Review by DavideIadeluca — APPROVED** [review]
Thank you for your patience. I had another look again and from my side this is good for a merge to master. Everything is working as expected locally.

About the merge conflicts, before we merge this to master I think the following should work to resolve the issue:

```
git fetch origin
git checkout origin/master -- js/dist/admin.js js/dist/admin.js.map js/dist/forum.js js/dist/forum.js.map
git commit -m "chore: resolve merge conflict"
git push origin feat/add-optional-policies
```

**Comment by rafaucau** [comment]
@DavideIadeluca merging issue resolved

**Comment by rafaucau** [comment]
@DavideIadeluca So what about merge?

**Comment by DavideIadeluca** [comment]
> @DavideIadeluca So what about merge?

Awaiting final approval from @imorland

**Comment by AngryDragonite** [comment]
ping @imorland

**Comment by AngryDragonite** [comment]
@imorland @DavideIadeluca is there any chance of getting this thing done in any near future? We've been waiting for 4 months.

**Comment by rafaucau** [comment]
@imorland

**Pull request merged**
